types/egg/index.d.ts

```diff
@@ -0,0 +1,765 @@
+   *
+   * when you set locals, only object is available
+   *
+   * ```js
+   * this.locals = {
+   *   a: 1
+   * };
+   * this.locals = {
+   *   b: 1
+   * };
+   * this.locals.c = 1;
+   * console.log(this.locals);
+   * {
+   *   a: 1,
+   *   b: 1,
+   *   c: 1,
+   * };
+   * ```
+   *
+   * `ctx.locals` has cache, it only merges `app.locals` once in one request.
+   *
+   * @member {Object} Context#locals
+   */
+  locals: any;
+
+  /**
+   * alias to {@link locals}, compatible with koa that use this variable
+   */
+  state: any;
+
+  /**
+   * Logger for Application, wrapping app.coreLogger with context infomation
+   *
+   * @member {ContextLogger} Context#logger
+   * @since 1.0.0
+   * @example
+   * ```js
+   * this.logger.info('some request data: %j', this.request.body);
+   * this.logger.warn('WARNING!!!!');
+   * ```
+   */
+  logger: Logger;
+
+  /**
+   * Request start time
+   */
+  starttime: number;
+
+  /**
+   * View instance that is created every request
+   */
+  view: ContextView;
+
+  /**
+   * http request helper base on httpclient, it will auto save httpclient log.
+   * Keep the same api with httpclient.request(url, args).
+   * See https://github.com/node-modules/urllib#api-doc for more details.
+   */
+  curl(url: string, opt: any): Promise<any>;
+
+  /**
+   * Render a file by view engine
+   * @param {String} name - the file path based on root
+   * @param {Object} [locals] - data used by template
+   * @param {Object} [options] - view options, you can use `options.viewEngine` to specify view engine
+   * @return {Promise<String>} result - return a promise with a render result
+   */
+  render(name: string, locals: any, options?: any): Promise<string>;
+
+  /**
+   * Render a template string by view engine
+   * @param {String} tpl - template string
+   * @param {Object} [locals] - data used by template
+   * @param {Object} [options] - view options, you can use `options.viewEngine` to specify view engine
+   * @return {Promise<String>} result - return a promise with a render result
+   */
+  renderString(name: string, locals: any, options?: any): Promise<string>;
+
+  __(key: string, ...values: string[]): string;
+  gettext(key: string, ...values: string[]): string;
+
+  /**
+   * get upload file stream
+   * @example
+   * ```js
+   * const stream = yield this.getFileStream();
+   * // get other fields
+   * console.log(stream.fields);
+   * ```
+   * @method Context#getFileStream
+   * @return {ReadStream} stream
+   * @since 1.0.0
+   */
+  getFileStream(): Promise<FileStream>;
+
+  /**
+   * @see Responce.redirect
+   */
+  redirect(url: string, alt?: string): void;
+}
+
+export class Controller extends BaseContextClass { }
+
+export class Service extends BaseContextClass { }
+
+export interface service {} // tslint:disable-line
```

Comment:
Having an interface with a lowercased version of a class' name is confusing, could this be renamed? We should start linting for upper-case type names.
This could also use a doc comment explaining how to use it.